src/main/java/edu/harvard/iq/dataverse/pidproviders/PidProviderFactoryBean.java

```diff
@@ -0,0 +1,230 @@
+        if (providerPathSetting.isPresent()) {
+            Path exporterDir = Paths.get(providerPathSetting.get());
+            // Get all JAR files from the configured directory
+            try (DirectoryStream<Path> stream = Files.newDirectoryStream(exporterDir, "*.jar")) {
+                // Using the foreach loop here to enable catching the URI/URL exceptions
+                for (Path path : stream) {
+                    logger.log(Level.FINE, "Adding {0}", path.toUri().toURL());
+                    // This is the syntax required to indicate a jar file from which classes should
+                    // be loaded (versus a class file).
+                    jarUrls.add(new URL("jar:" + path.toUri().toURL() + "!/"));
+                }
+            } catch (IOException e) {
+                logger.warning("Problem accessing external Exporters: " + e.getLocalizedMessage());
+            }
+        }
+        URLClassLoader cl = URLClassLoader.newInstance(jarUrls.toArray(new URL[0]), this.getClass().getClassLoader());
+
+        /*
+         * Step 2 - load all PidProcviderFactories that can be found, using the jars as
+         * additional sources
+         */
+        loader = ServiceLoader.load(PidProviderFactory.class, cl);
+        /*
+         * Step 3 - Fill pidProviderFactoryMap with type as the key, allow external
+         * factories to replace internal ones for the same type. FWIW: From the logging
+         * it appears that ServiceLoader returns classes in ~ alphabetical order rather
+         * than by class loader, so internal classes handling a given providerName may
+         * be processed before or after external ones.
+         */
+        loader.forEach(providerFactory -> {
+            String type = providerFactory.getType();
+            logger.fine("Loaded PidProviderFactory of type: " + type);
+            // If no entry for this providerName yet or if it is an external exporter
+            if (!pidProviderFactoryMap.containsKey(type) || providerFactory.getClass().getClassLoader().equals(cl)) {
+                logger.fine("Adding PidProviderFactory of type: " + type + " to the map");
+                pidProviderFactoryMap.put(type, providerFactory);
+            }
+            logger.log(Level.FINE,
+                    "Loaded PidProviderFactory of type: " + type + " from "
+                            + providerFactory.getClass().getCanonicalName() + " and classloader: "
+                            + providerFactory.getClass().getClassLoader().getClass().getCanonicalName());
+        });
+    }
+
+    private void loadProviders() {
+        try {
+            String[] providers = JvmSettings.PID_PROVIDERS.lookup(String[].class);
+            for (String name : providers) {
+                String type = JvmSettings.PID_PROVIDER_TYPE.lookup(name);
+                if (pidProviderFactoryMap.containsKey(type)) {

```

Comment:
The `else` branch is missing. We rely on string matching here: a typo in the type would render the configuration useless without letting the admin know. IMHO this is crucial and potentially should fail deployment or at least trigger a big fat warning.